### ISSUE TITLE
pyright not detecting that a list of optional do not contain None

### ISSUE THREAD

**Issue by fabiencro** [microsoft/pyright#489]
Hi,

It seems pyright cannot detect that a list of optional do not contain None. (I am using the pyright plugin of VS Code).

For example, in the following minimal example:

```python
from typing import Optional, List

def f(lst:List[Optional[int]]):
    assert all(a is not None for a in lst)
    b: List[int] = [a for a in lst]
```
pyright complains with the following message:
`Expression of type 'list[Optional[int]]' cannot be assigned to declared type 'List[int]'
  Type 'Optional[int]' cannot be assigned to type 'int'pyright`

Same problem with this version:

```python
def f(lst:List[Optional[int]]):
    for a in lst:
        assert a is not None
    b: List[int] = [a for a in lst]
```

It seems to me it should work, because pyright otherwise happily accept asserts to indicate an optional is not None. i.e. the following code is accepted:
```python
def g(k:Optional[int]):
    assert k is not None
    b:int = k
```

**Comment by erictraut** [comment]
Type checkers can't make assumptions about the inner workings of functions. They need to base their decisions on the information provided by type annotations and by documented behaviors of operations that are spec'ed as part of the Python standard.

In your first code snippet, Pyright has no idea what the function "all" does, so it can't make any decisions based on that assertion.

Likewise, with your second example. A type checker can't make assumptions about the implementation of the `__iter__` method for `lst` other than what the type stub indicates.

In general, type checkers will never narrow a _type argument_ based on an assertion or conditional statement. In the case of `List[Optional[int]]`, the type argument is `Optional[int]`. Note that `Optional[int]` is just another way of writing `Union[int, None]`. "Narrowing" is the process of eliminating one or more of the types in a union. So Pyright is able to narrow _types_ based on conditional statements and assertions, it will never narrow _type arguments_.

If you change your code to the following, Pyright will correctly infer that the type of `b` is `List[int]`.
```
b = [a for a in lst if a is not None]
```

**Comment by fabiencro** [comment]
Thank you for the very quick answer.

I get what you are saying. However it would not seem unreasonable to me to assume that an object typed with  the standard library `List` or `Sequence` has a "reasonable" __iter__ method and do inference based on this assumption ("reasonable" meaning for example that a straightforward for-loop will enumerate all objects in the container). 

But I understand this is not something you would want to  do in pyright.

And I found that the better way to handle this in my case is probably to use an explicit `typing.cast`:

```python
from typing import Optional, List, cast

def f(lst:List[Optional[int]]):
    b: List[int] = cast(List[int], [a for a in lst])
```

(Or just in case, do you have a better suggestion? I am a bit new to pyright and  python static typechecking, so I am not sure of best practices for all situations)

**Comment by erictraut** [comment]
A cast will work, but I generally use casts only as an absolute last resort. They turn off all checks in the type checker, making your code more fragile. If you're concerned about runtime overhead, then cast can be an effective solution. But in your case, you're using a cast _and_ paying the runtime overhead of iterating through all elements in your list, so the approach I suggested above would be better.

If you don't want to pay the runtime overhead, then it would be better to do the following:
```
b = cast(List[int], b)
```
(Note that the type annotation for `b` isn't required because you're already providing the type in the right-hand side in the cast function.)

**Comment by erictraut** [comment]
Regarding your point about hard-coding knowledge about built-in types... I understand your point, but the logic gets extremely complicated quickly. For example, determining what is a "straightforward for-loop" is very difficult because lists are mutable. So the type checker would need to track all possible mutations of a list in all possible loop constructs (including nested loops). There are lots of ways lists can be mutated, so all of those would need to be hard-coded into the logic of the type checker as well. And since Python supports subclassing of built-in types, all of the hard-coded assumptions about the built-in type would need to be revalidated (or assumed to be invalid) for subclasses. It's a very slippery slope to unbounded complexity.

**Comment by fabiencro** [comment]
Thanks a lot for your answer.

I do not like too much the `b = [a for a in lst if a is not None]` solution because I would rather avoid changing the semantic of the program just for typechecking (I am fine with asserts as they can easily be disabled at "production" time). So I guess I will use cast for now, even with all its drawbacks.

I do understand that you cannot handle correctly all situations. You would certainly end up with a mess of a turing-complete typechecker. However maybe supporting specific special-cases  like `assert all(  constraint(x) for x in lst)` would be doable and useful enough? It seems to me that the need to convert a container of optionals to a container of non-optionals is not so uncommon, and it is a bit annoying there is no clean way of doing it. But anyway, it might have more to do with python typechecking system in general and not pyright specifically.
